sdks/python/apache_beam/runners/portability/portable_runner.py

```diff
@@ -84,13 +84,12 @@ def default_docker_image():
                       'has Python %d.%d interpreter.' % (
                           sys.version_info[0], sys.version_info[1]))
 
-      # Perhaps also test if this was built?
-      image = ('{user}-docker-apache.bintray.io/beam/python'
-               '{version_suffix}:latest'.format(
-                   user=os.environ['USER'],
-                   version_suffix=version_suffix))
+      image = ('apachebeam/python{version_suffix}_sdk:latest'.format(
```

Comment:
We may need to give people a way to run with their locally built containers for testing purposes, or for using custom containers.

We could reuse worker_harness_container_image flag, introduce a new flag. Or perhaps something else to do this in an easy way.

@tweise might have suggestions.

Reply:
I have mixed feelings about this. When working with master, wouldn't I expect the locally built image to be used? What does it mean to refer to hub for something that wasn't released.

Perhaps do that depending on whether current version is dev/snapshot or a release?

Reply:
And thanks for the ping @aaltay !

Reply:
To build on Thomas's idea, we could do something similar to what dataflowrunner does today (https://github.com/apache/beam/blob/9043b3e0bc138e63123d3fa55d4b7280eb96a82e/sdks/python/apache_beam/runners/dataflow/internal/apiclient.py#L946)

- If this is a released version on release branch (i.e. no 'dev' in version') default to image in hub. (release manager will have to build this container as part of the release anyway.)
- If it is a dev version --> this is a question. Dataflowrunner's solution is to have a fixed image that is occasionally updated as needed or if provided a flag use that as an override. (Here we might just simplify and require a locally built version. This occasionally built image is convenient but is error-prone and a hassle.)

Reply:
Thanks @aaltay and @tweise for comments.

`--environment_config` option can be used to pass customized images when use Portable runner. The default image is used when no docker image is specified.

My understanding is, from users perspective, they don't need to worry about creating docker images at local when they write pipelines, so it's ok to pull it from remote. From developers perspective, developers who are working on sdks are expected to know how portable runner is working, so as how docker is used. If they want to test their developing sdks, they should build an image at local and use it. The image name can be same as the default one, then the local image is used instead of pulling it from remote, or can pass it with  `--environment_config` option if it has different name.

I can add more comments here to explain how to pass customized images.

What scenarios am I missing?

Reply:
When the Beam version is not a development version (a release), then by default use the image corresponding to that release. Not `latest`, because that could be anything and may not be compatible with the release version? If the Python Beam version is 2.16, then it should not pick up something that could be 2.17, for example.

If the version of Beam is a development version, continue to use the locally built image as before.

Reply:
But if a locally built container would automatically be picked up, then the distinction won't be necessary. Is that already the case with these changes?

Reply:
Ah, I see, it’s possible that Beam version is not up to date. 
It will check local first, if the image is not available, them pull it from remote.